arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp

```diff
@@ -101,103 +104,145 @@ auto DocumentStateTransactionHandler::getTrx(TransactionId tid)
-
-          LOG_CTX("0da00", INFO, logContext)
-              << "Result ignored while applying transaction " << doc.tid
-              << " with operation " << int(doc.operation) << " on shard "
-              << doc.shardId << ": " << res;
-          return Result{};
-        }
-        return res;
-      }
-      case OperationType::kCommit: {
-        auto res = trx->commit();
-        removeTransaction(doc.tid);
-        return res;
-      }
-      case OperationType::kAbort: {
-        auto res = trx->abort();
-        removeTransaction(doc.tid);
-        return res;
-      }
-      case OperationType::kAbortAllOngoingTrx:
-        TRI_ASSERT(false);  // should never happen as it should be handled above
-        FATAL_ERROR_EXIT();
-      case OperationType::kIntermediateCommit:
-        return trx->intermediateCommit();
-      default:
-        THROW_ARANGO_EXCEPTION(TRI_ERROR_TRANSACTION_DISALLOWED_OPERATION);
+void DocumentStateTransactionHandler::removeTransaction(TransactionId tid) {
+  _transactions.erase(tid);
+}
+
+auto DocumentStateTransactionHandler::getUnfinishedTransactions() const
+    -> TransactionMap const& {
+  return _transactions;
+}
+
+auto DocumentStateTransactionHandler::getTransactionsForShard(
+    ShardID const& sid) -> std::vector<TransactionId> {
+  std::vector<TransactionId> result;
+  for (auto const& [tid, trx] : _transactions) {
+    if (trx->containsShard(sid)) {
+      result.emplace_back(tid);
     }
-  } catch (basics::Exception& e) {
-    return Result{e.code(), e.message()};
-  } catch (std::exception& e) {
-    return Result{TRI_ERROR_TRANSACTION_INTERNAL, e.what()};
   }
+  return result;
 }
 
-auto DocumentStateTransactionHandler::ensureTransaction(
-    DocumentLogEntry const& doc) -> std::shared_ptr<IDocumentStateTransaction> {
-  TRI_ASSERT(doc.operation != OperationType::kAbortAllOngoingTrx);
+[[nodiscard]] auto DocumentStateTransactionHandler::validate(
+    ReplicatedOperation::OperationType const& operation) const -> Result {
+  return std::visit(
+      [&](auto&& op) -> Result {
+        using T = std::decay_t<decltype(op)>;
+        if constexpr (ModifiesUserTransaction<T>) {
+          if (!_shardHandler->isShardAvailable(op.shard)) {
+            return Result{TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND};
+          }
+        }
+        return Result{};
+      },
+      operation);
+}
 
-  auto tid = doc.tid;
-  auto trx = getTrx(tid);
-  if (trx != nullptr) {
-    return trx;
+auto DocumentStateTransactionHandler::applyOp(
+    FinishesUserTransaction auto const& op) -> Result {
+  TRI_ASSERT(op.tid.isFollowerTransactionId());
+  auto trx = getTrx(op.tid);
+  ADB_PROD_ASSERT(trx != nullptr);
+
+  auto res = Result{};
+  using T = std::decay_t<decltype(op)>;
+  if constexpr (std::is_same_v<T, ReplicatedOperation::Commit>) {
+    res = trx->commit();
+  } else if constexpr (std::is_same_v<T, ReplicatedOperation::Abort>) {
+    res = trx->abort();
   }
```

Comment:
I would prefer something like this:
```suggestion
  auto res = std::invoke(
    overload{
      [&](ReplicatedOperation::Commit const&) { return trx->commit(); },
      [&](ReplicatedOperation::Abort const&) { return trx->abort); }
    }, op);
```
I think this is a bit easier to read and it would also cause a compile error if we ever add a new type to `FinishesUserTransaction`.

Reply:
I updated the method.